modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointReadWriteLock.java

```diff
@@ -182,14 +199,35 @@ public boolean hasQueuedWriters() {
     }
 
     private void onReadLock(long start, boolean taken) {
-        long elapsed = coarseCurrentTimeMillis() - start;
+        readLockWaitingThreadsGauge.decrementAndGet();
+
+        long coarseCurrentTimeMillis = coarseCurrentTimeMillis();
+        long elapsed = coarseCurrentTimeMillis - start;
 
         if (taken) {
+            if (checkpointReadLockAcquiredTime.get() == null) {
+                checkpointReadLockAcquiredTime.set(coarseCurrentTimeMillis);
+            }
             checkpointReadLockHoldCount.set(checkpointReadLockHoldCount.get() + 1);
+            metrics.recordReadLockAcquisitionTime(elapsed);
         }
 
         if (elapsed > LONG_LOCK_THRESHOLD_MILLIS) {
             log.warn(LONG_LOCK_THROTTLE_KEY, "Checkpoint read lock took {} ms to acquire.", elapsed);
         }
     }
+
+    private void onReadUnlock() {
+        int previousLockCount = checkpointReadLockHoldCount.get();
+        checkpointReadLockHoldCount.set(previousLockCount - 1);
+        if (previousLockCount == 1) {
+            // fully unlocked
+            Long checkpointReadLockAcquiredTime = this.checkpointReadLockAcquiredTime.get();
+            if (checkpointReadLockAcquiredTime != null) {
+                this.checkpointReadLockAcquiredTime.remove();
```

Comment:
Removing the thread local and then reinitializing it is not free, I'd suggest not doing it without strong reasons